### PR TITLE
Create privatelink-with-network-firewall.yaml

### PR DESCRIPTION
**Pull request by niroy**

### CODE REVIEW

File: privatelink-with-network-firewall.yaml [hunk 1]
```diff
@@ -0,0 +1,746 @@
...
+  rEgressFirewallRuleGroup:
+    Type: "AWS::NetworkFirewall::RuleGroup"
+    Properties:
+      Capacity: "100"
+      RuleGroupName: !Join ["", [!Ref pRosaClusterName, "-firewall-rule-group"]]
+      Type: "STATEFUL"
+      Tags:
+        - Key: "Name"
+          Value: !Join ["", [!Ref pRosaClusterName, "-firewall-rule-group"]]
+      RuleGroup:
```
Comment (herve4m) [inline privatelink-with-network-firewall.yaml:328]: I think you also need to set the `HOME_NET` variable to filter traffic from the ROSA VPC.
See https://docs.aws.amazon.com/network-firewall/latest/developerguide/stateful-rule-groups-domain-names.html#stateful-rule-groups-domain-names-home-net

```suggestion
      RuleGroup:
        RuleVariables:
          IPSets:
            HOME_NET:
              Definition:
                - "10.0.0.0/16"
                - "10.1.0.0/16"
```

File: privatelink-with-network-firewall.yaml [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+      FirewallPolicyName:
+        !Join ["", [!Ref pRosaClusterName, "-firewall-policy"]]
+      FirewallPolicy:
+        StatelessDefaultActions:
+          - "aws:forward_to_sfe"
+        StatelessFragmentDefaultActions:
+          - "aws:forward_to_sfe"
+        StatefulRuleGroupReferences:
+          - ResourceArn: !Ref rEgressFirewallRuleGroup
+      Description: Stateful rule group for Network Firewall
```
Comment (herve4m) [inline privatelink-with-network-firewall.yaml:317]: I would also add a tag. Maybe:

```suggestion
      Description: Stateful rule group for Network Firewall
      Tags:
        - Key: "Name"
          Value: !Join ["", [!Ref pRosaClusterName, "-network-firewall-policy"]]
```

**Comment by herve4m** [comment]
1. (filter by path), you have to allow the hostname (cloud.redhat.com): you cannot filter by path because it's 443 (TLS) so the firewall cannot inspect the URL.

2. For port 9997, it seems that the URLs (inputs1.osdsecuritylogs.splunkcloud.com for example) use TLS (but not HTTPS):

```
$ openssl s_client -connect inputs1.osdsecuritylogs.splunkcloud.com:9997
CONNECTED(00000003)
depth=2 CN = Root CA
verify error:num=19:self-signed certificate in certificate chain
verify return:1
depth=2 CN = Root CA
verify return:1
depth=1 CN = Intermediate CA
verify return:1
depth=0 CN = *.osdsecuritylogs.splunkcloud.com
verify return:1
804B7223507F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1605:SSL alert number 40
---
Certificate chain
 0 s:CN = *.osdsecuritylogs.splunkcloud.com
   i:CN = Intermediate CA
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Oct  3 01:11:44 2022 GMT; NotAfter: Oct  2 01:12:14 2027 GMT
 1 s:CN = Intermediate CA
   i:CN = Root CA
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Oct  3 22:50:18 2019 GMT; NotAfter: Sep 30 22:50:48 2029 GMT
 2 s:CN = Root CA
   i:CN = Root CA
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Oct  3 04:58:37 2019 GMT; NotAfter: Sep 30 04:59:01 2029 GMT
---
...
```

You have to use suricata rules (called signatures) in a AWS::NetworkFirewall::RuleGroup > Properties > RuleGroup > RulesSource > RulesString.
Suricata doc: https://docs.suricata.io/en/latest/rules/index.html
For that case, use the `tls.cert_subject` filter. Something such as:

```
pass tls $HOME_NET any -> $EXTERNAL_NET 9997 (tls.cert_subject; dotprefix; content:".osdsecuritylogs.splunkcloud.com"; endswith; msg:"Pass TLS to .osdsecuritylogs.splunkcloud.com"; sid:1004;  rev:1;)
```

Example on adding such a rule in CloudFormation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-networkfirewall-rulegroup.html#aws-resource-networkfirewall-rulegroup--examples


For port 22, get the IP address of sftp.access.redhat.com and filter (suricata rule or stateless rule) by IP.

**Comment by herve4m** [comment]
You also need to connect the firewall. I think you need to add another private subnet that hosts the firewall (between rEgressVpcPrivateSubnet and rEgressVpcPublicSubnet), add a route table, and update the other route tables.
Example: https://medium.com/slalom-technology/secure-internet-access-egress-filtering-with-aws-network-firewall-ddf52ae121f9

File: privatelink-with-network-firewall.yaml (hunk 1 above)
Comment (niroy) [reply privatelink-with-network-firewall.yaml:328]: Thanks, added

File: privatelink-with-network-firewall.yaml (hunk 2 above)
Comment (niroy) [reply privatelink-with-network-firewall.yaml:317]: Added tag

File: privatelink-with-network-firewall.yaml [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+                - "10.0.0.0/16"
+                - "10.1.0.0/16"
+        RulesSource:
+          RulesSourceList:
+            TargetTypes:
+              - "TLS_SNI"
+              - "HTTP_HOST"
+            Targets:
+              - "registry.redhat.io"
+              - ".quay.io"
```
Comment (herve4m) [inline privatelink-with-network-firewall.yaml:349]: Missing `quay.io` (in addition to `.quay.io`):
```suggestion
              - ".quay.io"
              - "quay.io"
```

**Comment by niroy** [comment]
> You also need to connect the firewall. I think you need to add another private subnet that hosts the firewall (between rEgressVpcPrivateSubnet and rEgressVpcPublicSubnet), add a route table, and update the other route tables. Example: https://medium.com/slalom-technology/secure-internet-access-egress-filtering-with-aws-network-firewall-ddf52ae121f9

Yes, for this one I found this guide: https://docs.aws.amazon.com/whitepapers/latest/building-scalable-secure-multi-vpc-network-infrastructure/using-nat-gateway-with-firewall.html
While, this perfectly fits our requirement, I'm still trying to figure out the routing table updates needed. Mainly confusion is on Transit gateway to egress vpc route and the blackhole route. I think we should discuss some more on this before we proceed

**Comment by herve4m** [comment]
> 
 I've created a diagram with the current (non-working) template and one possible solution:

https://drive.google.com/file/d/18HfyFNGH3IU43SC4I6J4tltdjUBakz8R/view

We can use it to discuss and comment.

You're right, it's tricky if we go with blackhole routes (but that increases security between connected VPCs). Let's research on that.

File: privatelink-with-network-firewall.yaml (hunk 3 above)
Comment (niroy) [reply privatelink-with-network-firewall.yaml:349]: This gives error if I include quay.io 
`Resource handler returned message: "Targets has duplicate values, parameter: [quay.io], context: RulesSource.RulesSourceList.Targets[2] (Service: NetworkFirewall, Status Code: 400, Request ID: 8dabff27-e8e4-40fa-aba4-e5928bddc0d0)" (RequestToken: b189b206-f146-e004-56a9-1a9d75ac1dc9, HandlerErrorCode: InvalidRequest)`

Comment (herve4m) [reply privatelink-with-network-firewall.yaml:351]: Fixed

**Review by herve4m — APPROVED** [review]

**Pull request merged**